### PR TITLE
Parameterize the use of latex engine

### PR DESCRIPTION
**Pull request by solsjo**
This pull request aims at simplifying the rules, raise the readability, reduce the build time of the builds and promote extensibility.

Build times are reduced by not copying all binaries to a structure similar to the one latex expects.
Latex / kpathsea assumes that directories with only symlinks to be empty.
This is overcome by instead setting the environment variables that kpathsea makes use of
(kpathsea is the lib responsible for locating files in the tex environment).

Apart from this further refactoring might be able to do away with the tool_wrapper altogether.

### CODE REVIEW

**Comment by solsjo** [comment]
it seems setting the environment variables from starlark results in a lot of memory being claimed. It might be wise to handle the setting via Args instead, so as to more efficiently handle it.
Or pass the envs, as an action parameter instead of as an argument.

**Comment by solsjo** [comment]
@Kernald Can you have a look at this pull request?

**Comment by solsjo** [comment]
@patbro would you have time to review this pull request?

**Comment by Kernald** [comment]
Sorry for the delay - I'll try looking at these over the next few days!

File: latex.bzl [hunk 1]
```diff
@@ -71,20 +54,38 @@ _latex = rule(
...
         "main": attr.label(
             allow_single_file = [".tex"],
             mandatory = True,
         ),
         "srcs": attr.label_list(allow_files = True),
+        "web2c": attr.label(
+            allow_files = True,
+            default = "@texlive_texmf__texmf-dist__web2c",
+        ),
+        #       the user can decide his or hers engine of choice
```
Comment (Kernald) [inline latex.bzl]: Nit: `their` would be simpler :-)

File: latex_engine_cmds.bzl [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    In general latex distributions are encouraged to follow the 'TDS'
+    structure. And tools and script might make assumptions that the
+    layout of directories respects that structure.
+
+    But from our perspective, trying to shoehorn the partioned repositories
+    according
+    (E.g. <sandbox>/bazel-latex/external/<ext_texlive_repo>/<files>)
+    to the TDS creates an unnecessary complexity.
+    Also kpathsea tries to be efficient about looking up files.
+    So to derive if a folder is of interest is check if the number
```
Comment (Kernald) [inline latex_engine_cmds.bzl]: This line is a bit unclear - do you mean `So to derive if a folder is of interest, we check...` maybe? The `is check` seems wrong, but I'm not sure what it should actually be.

File: latex_engine_cmds.bzl [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    of files or folders in the current folder being inspected is greater than 2.
```
Comment (Kernald) [inline latex_engine_cmds.bzl]: Tiny nit: this line (and a few others) is too long.

File: latex_engine_cmds.bzl [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+      latex_tool: The latex engine to use, in this case we make the
+        assumption that it is the luahbtex binary (This should be
+        enforced more strongly).
+
+    Returns:
+      A list with the three commands
+    """
+
+    # TODO: Each command could be its own invokation to ctx.actions.run
+    #       this way the artifacts could be cache, it might not help that
```
Comment (Kernald) [inline latex_engine_cmds.bzl]: Nit: `could be cached`

File: latex_engine_cmds.bzl [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    #       much though, as any change to the inputs is likely to re-trigger
+    #       a re-build.
```
Comment (Kernald) [inline latex_engine_cmds.bzl:230]: Regarding rebuilds: I'm not too familiar with the inner workings of LaTeX compilers, but if the first command generates an intermediate representation of sorts, it definitely would make sense to separate the different commands - if someone fixes e.g. indentation in a source file, any intermediate representation should likely be the same, hence avoiding future steps.

(Note that I haven't read the implementation yet...)

File: texlive_2022_repos.bzl [hunk 6]
```diff
@@ -0,0 +1,6591 @@
+"""
```
Comment (Kernald) [inline texlive_2022_repos.bzl:1]: I'm a bit surprised this isn't the same length as `repositories.bzl` was - I was assuming you renamed it. Is there any way we could preserve the history here, to see what actually changed, by maybe postponing the renaming?

File: tool_wrapper.py [hunk 7]
```diff
@@ -63,74 +23,77 @@ def setup_argparse():
...
+        env = setup_env(args.env)
+        cmd_args = [args.tool] + args.flag + [args.input]
+        returncode = 0
+        output = None
+        res = ""
+        if not args.input:
+            cmd_args = cmd_args[:-1]
+
+        cwd="."
+        shell = False
```
Comment (Kernald) [inline tool_wrapper.py]: I'm not sure keeping `cwd` and `shell` declared independently from the call just below brings much value.

**Review by Kernald — CHANGES_REQUESTED** [review]

File: latex_engine_cmds.bzl (hunk 2 above)
Comment (solsjo) [reply latex_engine_cmds.bzl]: Yeah, I noticed that the sentence was weird after I had pushed it. I'll fix it!

File: latex_engine_cmds.bzl (hunk 3 above)
Comment (solsjo) [reply latex_engine_cmds.bzl]: Ack, will fix!

File: texlive_2022_repos.bzl (hunk 6 above)
Comment (solsjo) [reply texlive_2022_repos.bzl:1]: Ahh, good point!
Sure, I can move out the logic to a separate file first. Then do a rename?

File: tool_wrapper.py (hunk 7 above)
Comment (solsjo) [reply tool_wrapper.py]: No, this was probably a remains from debugging. Will fix.

**Comment by solsjo** [comment]
No worries! Thanks for the review @Kernald. I have a follow up commit that removes the tool_wrapper altogether as well. When you have time :)

File: latex.bzl (hunk 1 above)
Comment (solsjo) [reply latex.bzl]: done

File: latex_engine_cmds.bzl (hunk 2 above)
Comment (solsjo) [reply latex_engine_cmds.bzl]: done

File: latex_engine_cmds.bzl (hunk 3 above)
Comment (solsjo) [reply latex_engine_cmds.bzl]: done

File: latex_engine_cmds.bzl (hunk 4 above)
Comment (solsjo) [reply latex_engine_cmds.bzl]: done

File: latex_engine_cmds.bzl (hunk 5 above)
Comment (solsjo) [reply latex_engine_cmds.bzl:230]: Something along those lines.
No I'm not well versed in the inner workings of latex either. We should probably ask someone from the latex community that knows their way around the tex distro code to review this code. I feel like we're aiming in the dark.

File: tool_wrapper.py (hunk 7 above)
Comment (solsjo) [reply tool_wrapper.py]: done

**Comment by solsjo** [comment]
Updated according to the comments.

**Comment by solsjo** [comment]
@Kernald

File: texlive_2022_repos.bzl (hunk 6 above)
Comment (solsjo) [reply texlive_2022_repos.bzl:1]: done

**Review by Kernald — APPROVED** [review]

**Comment by Kernald** [comment]
Thanks!

**Pull request merged**